### ISSUE TITLE
WebSocket requestParameterMap not being populated

### ISSUE THREAD

**Issue by japgolly** [eclipse/jetty.project#3575]
Given this minimal WebSocket and custom Configurator:

```scala
import javax.websocket._
import javax.websocket.server._

final class Connector extends ServerEndpointConfig.Configurator {
  override def modifyHandshake(cfg: ServerEndpointConfig, req: HandshakeRequest, res: HandshakeResponse): Unit = {
    println("[modifyHandshake] params = " + req.getParameterMap.keySet())
  }
}

@ServerEndpoint(value = "/w/p/{hello}", configurator = classOf[Connector])
final class WebSocket {
  @OnOpen
  def onOpen(s: Session, @PathParam("hello") hello: String): Unit = {
    println("[onOpen] hello = " + hello)
    println("[onOpen] params = " + s.getRequestParameterMap.keySet())
  }
}
```

If I connect to `ws://localhost:8080/w/p/hmmm` I see the following output:

```
[modifyHandshake] params = []
[onOpen] hello = hmmm
[onOpen] params = []
```

The problem is that the the requestParameterMap is empty, where as I'm expecting that it should be a map with a single key of `hello`. I'm basing that expectation from my understanding of the JavaDoc, and apparently-working examples in the wild ([[1]](https://stackoverflow.com/questions/53405281/websocket-pathparam-validation) [[2]](https://stackoverflow.com/questions/46009847/how-to-properly-report-an-error-to-client-through-websockets)).

Jetty version `9.4.16.v20190411` running on `OpenJDK 64-Bit Server VM, Java 1.8.0_212`

**Comment by japgolly** [comment]
Oh and I forgot to mention: the reason I want to access params programmatically without using the `@PathParam` annotation, is that I need access in the `ServerEndpointConfig.Configurator` in which one doesn't allow for annotations.

**Comment by lachlan-roberts** [comment]
@japgolly The `@PathParam` annotation is giving path params, but `session.getRequestParameterMap()` actually gets a map of parameters from the URL query string.

So if you connected to `ws://localhost:8080/w/p/hmmm?foo=bar`, you would see values in the RequestParameterMap.

Try using `Session.getPathParameters()`, this should give you the map of path params.

**Comment by japgolly** [comment]
Thanks @lachlan-roberts, I'll give that a go this week.

What do you think about reports of other people being able to access path params via `getRequestParameterMap`? Is it just undefined behaviour and other engines happen to behave differently?

**Comment by joakime** [comment]
@japgolly the `javax.websocket.Session` has 2 API methods.

* [`.getPathParameters()`](https://docs.oracle.com/javaee/7/api/javax/websocket/Session.html#getPathParameters--) - this is for "path" parameters, ones that exist on the supplied path.  This is almost exclusively used for returning the values associated with the a URI template and the usage of `@PathParam` annotation. (it should be noted that using `/a/path;foo/to;bar/a;baz/endpoint` is also a form of path parameter supported by many web containers, including jetty, not just ones using `javax.websocket`)
* [`.getRequestParameterMap()`](https://docs.oracle.com/javaee/7/api/javax/websocket/Session.html#getRequestParameterMap--) - this is for "request parameters", aka [`HttpServletRequest.getParameterMap()`](https://docs.oracle.com/javaee/7/api/javax/servlet/ServletRequest.html#getParameterMap--). (notice the similar method interface?).  This is for things like query parameters (and in the case of the servlet spec the POST body content parameters from `application/x-www-form-urlencoded` or `multipart/form-data` supplied content.   but body content on HTTP/1.1 websocket upgrade isn't supported by the spec, or jetty's implementation)

It's been this way since Jetty 9.0.0 (released in late 2012)

**Comment by japgolly** [comment]
Sorry for the delay. I've spent some more time on this and I think there's still a problem.

If one can get their hands on an instance of `javax.websocket.Session`, then calling `getPathParameters` looks to be great but the problem is that there doesn't seem to be a way to get a `javax.websocket.Session` instance from within a `ServerEndpointConfig.Configurator`. In `modifyHandshake` you're provided an instance of `HandshakeRequest` which has `getHttpSession()` but that returns a `org.eclipse.jetty.server.session.Session` which doesn't seem to have access to a `javax.websocket.Session` either.

In other words, accessing path params in a web socket instance is fine but how does one do so from a `ServerEndpointConfig.Configurator` instance?

**Comment by lachlan-roberts** [comment]
I don't think you can access the `Session` at this point

you could wait until `onOpen` is called to do the logic

or you could try parsing the path parameters yourself by doing something like this
```java
private UriTemplatePathSpec pathSpec = new UriTemplatePathSpec("/test/{foo}");

public Map<String, String> getPathParams(String path)
{
    Map<String, String> pathParamMap = pathSpec.getPathParams(path);
    return pathParamMap == null ? new HashMap<>() : pathParamMap;
}

public void modifyHandshake(ServerEndpointConfig sec, HandshakeRequest request, HandshakeResponse response)
{
    String uriString = request.getRequestURI().toString();
    System.err.println(getPathParams(uriString));
}
```
